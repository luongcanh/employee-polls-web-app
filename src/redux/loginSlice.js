import { createSlice } from '@reduxjs/toolkit';

const initialState = {};

export const loginSlice = createSlice({
	name: 'login',
	initialState,
	reducers: {
		setLogin: (state, action) => {
			state.authedUser = action.payload.id;
		},
		setLogout: (state) => {
			state.authedUser = null;
		}
	},
});

export const { setLogin, setLogout } = loginSlice.actions;

export const selectLogin = (state) => state?.login?.authedUser;

export default loginSlice.reducer;
