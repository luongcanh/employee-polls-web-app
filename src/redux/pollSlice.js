import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getAllPolls } from './utils/callAPI';
import { ANSWER_OPTION } from '../constants/constant';
import { checkIfAnswered } from './utils/common';
import { _saveQuestion, _saveQuestionAnswer } from '../_DATA';

const initialState = {
	loading: false,
	polls: [],
};

/**
 * @description Get all polls
 */
export const fetchPolls = createAsyncThunk('polls/getAllPolls', async () => await getAllPolls());

export const handleSubmitAnswer = createAsyncThunk(
	'polls/handleSubmitAnswer',
	async ({ authedUser, qid, answer }) => {
		await _saveQuestionAnswer({ authedUser, qid, answer });

		return { authedUser, qid, answer };
	},
);

/**
 * @description Create new poll
 */
export const handleCreateQuestion = createAsyncThunk(
	'polls/handleCreateQuestion',
	async ({ author, optionOneText, optionTwoText }) => {
		const result = await _saveQuestion({ author, optionOneText, optionTwoText });

		return result;
	},
);

const pollSlice = createSlice({
	name: 'polls',
	initialState,
	reducers: {},
	extraReducers: (builder) => {
		builder.addCase(fetchPolls.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(fetchPolls.fulfilled, (state, action) => {
			state.polls = action.payload;
			state.loading = false;
		});
		builder.addCase(handleSubmitAnswer.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(handleSubmitAnswer.fulfilled, (state, action) => {
			let currentVoters = state.polls[action.payload.qid][action.payload.answer].votes;
			if (
				!checkIfAnswered(state.polls[action.payload.qid], action.payload.authedUser) &&
				(action.payload.answer === ANSWER_OPTION.OPTION_ONE ||
					action.payload.answer === ANSWER_OPTION.OPTION_TWO)
			) {
				currentVoters.push(action.payload.authedUser);
			}
			state.loading = false;
		});
		builder.addCase(handleCreateQuestion.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(handleCreateQuestion.fulfilled, (state, action) => {
			state.loading = false;
			state.polls[action.payload.id] = action.payload;
		});
		builder.addCase(handleCreateQuestion.rejected, (state, action) => {
			state.loading = false;
		});
	},
});

export default pollSlice.reducer;
