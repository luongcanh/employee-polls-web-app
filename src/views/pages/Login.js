import { useDispatch } from 'react-redux';
import LoginIcon from '@mui/icons-material/Login';

import * as React from 'react';
import {
	CssBaseline,
	TextField,
	Box,
	Grid,
	Typography,
	Snackbar,
	createTheme,
	ThemeProvider,
	Alert
} from '@mui/material';
import { useLocation, useNavigate } from 'react-router-dom';
import { setLogin } from '../../redux/loginSlice';
import { checkLogin } from '../../redux/utils/callAPI';
import { LoadingButton } from '@mui/lab';

const defaultTheme = createTheme();

const Login = () => {
	const location = useLocation();
	const [showError, setShowError] = React.useState(false);
	const dispatch = useDispatch();
	const navigate = useNavigate();

	/**
	 * @description Handle the "Sign In" button click event
	 */
	const handleSubmit = (event) => {
		event.preventDefault();
		const data = new FormData(event.currentTarget);
		const usernameVal = data.get('username');
		const passwordVal = data.get('password');
		checkLogin(usernameVal, passwordVal).then((result) => {
			if (result) {
				dispatch(setLogin({ id: usernameVal }));
				navigate(location?.state?.prevUrl ? location?.state?.prevUrl : '/');
			} else {
				setShowError(true);
			}
		});
	};

	/**
 	* @description Close the error message
 	*/
	const handleCloseAlert = () => {
		setShowError(false)
	}

	return (
		<ThemeProvider theme={defaultTheme}>
			<Grid container component="main" sx={{ height: '100vh' }}>
				<CssBaseline />
				<Box sx={{ my: 8, mx: 'auto', display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'center', height: '100%', }} >
					<Typography component="h1" variant="h5">
						Sign in
					</Typography>
					<Box component="form" onSubmit={handleSubmit} sx={{ mt: 1 }}>
						<TextField margin="normal" required fullWidth id="username" label="Username" name="username" autoFocus data-testid='username' />
						<TextField margin="normal" required fullWidth name="password" label="Password" type="password" data-testid='password' />
						<LoadingButton type="submit" fullWidth variant="contained" sx={{ mt: 3, mb: 2 }} data-testid='submit-btn' startIcon={<LoginIcon />}>
							Sign In
						</LoadingButton>
						<Snackbar autoHideDuration={3000} anchorOrigin={{ vertical: 'top', horizontal: 'right' }} open={showError} onClose={handleCloseAlert} data-testid='error-snackbar'>
							<Alert severity="error">The username or password is incorrect. Please try again!</Alert>
						</Snackbar>
					</Box>
				</Box>
			</Grid>
		</ThemeProvider>
	);
};

export default Login;
