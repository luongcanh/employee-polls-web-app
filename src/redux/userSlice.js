import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getAllUsers } from './utils/callAPI';
import { handleCreateQuestion, handleSubmitAnswer } from './pollSlice';

const initialState = {
    loading: false,
    users: [],
};

export const fetchUsers = createAsyncThunk('users/getAllUsers', async () => await getAllUsers());

const userSlice = createSlice({
    name: 'users',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(fetchUsers.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(fetchUsers.fulfilled, (state, action) => {
            state.users = action.payload;
            state.loading = false;
        });
        builder.addCase(handleSubmitAnswer.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(handleSubmitAnswer.fulfilled, (state, action) => {
            state.users[action.payload.authedUser].answers[action.payload.qid] = action.payload.answer;
            state.loading = false;
        });
        builder.addCase(handleCreateQuestion.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(handleCreateQuestion.fulfilled, (state, action) => {
            state.loading = false;
            state.users[action.payload.author].questions.push(action.payload.id);
        });
        builder.addCase(handleCreateQuestion.rejected, (state, action) => {
            state.loading = false;
        });
    },
});

export default userSlice.reducer;
