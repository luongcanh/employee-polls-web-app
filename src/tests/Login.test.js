/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import Login from '../views/pages/Login';
import { store } from '../redux/store';


describe('Login Component', () => {
	let component;

	beforeEach(() => {

		// eslint-disable-next-line
		component = render(
			<Provider store={store}>
				<Router>
					<Login />
				</Router>
			</Provider>,
		);
	});

	it('render the login component correctly', () => {
		const { getByTestId } = component;

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
		expect(getByTestId('submit-btn')).toBeInTheDocument();
		expect(getByTestId('username')).toBeInTheDocument();
		expect(getByTestId('password')).toBeInTheDocument();
	});

	it('completes the form submission using correct credentials', async () => {
		const { getByTestId } = component;

		fireEvent.change(getByTestId('username').querySelector('input'), { target: { value: 'sarahedo' }, });
		fireEvent.change(getByTestId('password').querySelector('input'), { target: { value: 'password123' }, });
		fireEvent.click(getByTestId('submit-btn'));

		await waitFor(() => expect(store.getState().login.authedUser) === 'sarahedo');
	});
});
