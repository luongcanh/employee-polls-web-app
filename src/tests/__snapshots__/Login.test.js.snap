// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component render the login component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="MuiGrid-root MuiGrid-container css-1jrn2fe-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-vnisl2"
        >
          <h1
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            Sign in
          </h1>
          <form
            class="MuiBox-root css-164r41r"
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              data-testid="username"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="username"
                id="username-label"
              >
                Username
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="username"
                  name="username"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      Username
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              data-testid="password"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r1:"
                id=":r1:-label"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Password
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1qudo6r-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              data-testid="submit-btn"
              id=":r2:"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="LoginIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 7 9.6 8.4l2.6 2.6H2v2h10.2l-2.6 2.6L11 17l5-5zm9 12h-8v2h8c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-8v2h8z"
                  />
                </svg>
              </span>
              Sign In
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </form>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="MuiGrid-root MuiGrid-container css-1jrn2fe-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-vnisl2"
      >
        <h1
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Sign in
        </h1>
        <form
          class="MuiBox-root css-164r41r"
        >
          <div
            class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            data-testid="username"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="username"
              id="username-label"
            >
              Username
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="username"
                name="username"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Username
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            data-testid="password"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r1:"
              id=":r1:-label"
            >
              Password
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                name="password"
                required=""
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Password
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1qudo6r-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
            data-testid="submit-btn"
            id=":r2:"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="LoginIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 7 9.6 8.4l2.6 2.6H2v2h10.2l-2.6 2.6L11 17l5-5zm9 12h-8v2h8c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-8v2h8z"
                />
              </svg>
            </span>
            Sign In
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
